frontend/src/app/_services/schedule.service.ts

```diff
@@ -239,10 +239,48 @@ export interface Initial {
 }
 
 export interface GamingMeta {
-    id: string;
-    loc: string[];
+  id: string;
+  loc: string[];
+}
+
+/** Takes a set of dateRanges sorted by start time. */
+function filterEventsByDate(events: ScheduleEvent[], dateRanges: DateRange[], count?: number): ScheduleEvent[] {
+  if (count === 0) {
+    return [];
+  }
+  let i = 0;
+  const eventCount = events.length;
+  let filteredEvents: ScheduleEvent[] = [];
+  for (const dateRange of dateRanges) {
+    for (; i < eventCount && (!count || filteredEvents.length < count) && events[i].start < dateRange.start; i++) {
+      if (dateRange.inclusive && events[i].start.getTime() + events[i].mins * 60 * 1000 >= dateRange.start.getTime()) {
+        filteredEvents.push(events[i]);
+      }
+    }
+    if (!dateRange.end) {
+      filteredEvents = filteredEvents.concat(events.slice(i, count ? i + count - filteredEvents.length : undefined));
+      break;
+    }
+    for (; i < eventCount && (!count || filteredEvents.length < count) && events[i].start < dateRange.end; i++) {
+      filteredEvents.push(events[i]);
+    }
```

Comment:
having trouble mentally parsing this so i'm going to assume it works and move on